KnowledgeExtractor/WikipediaKnowledgeGraphExtractor.cs

```diff
@@ -96,34 +101,70 @@ public static KnowledgeGraph ExtractKnGraphFromUris(List<UriAndOriginalGraphType
-            // get the word counts from the json file
-            List<WordCount> dsWordsForGraph = JsonConvert.DeserializeObject<List<WordCount>>(File.ReadAllText("../../../dataStructureWordsCountForNodesInGraph.json"));
-            // get the nodes for the DS words
-            List<KnGNode> dataStructureNodes = result.KnGraph.Where(n => n.OriginalGraphType == OriginalGraphType.DataStructuresKnGraph).ToList();
-            List<string> dsWordsToIgnore = new List<string>() { "other", "union", "reference", "arrays", "image", "matrix", "integer", "character", "list", "stack", "queue", "record", "bitmap", "set" };
-            // add the additional edges to the graph according to the json file
-            foreach (var dsWordsForNode in dsWordsForGraph)
+        private static void AddDataStructureWordsToNode(ref KnowledgeGraph graph, WordCount wordsToAddToNode, Dictionary<string, KnGNode> dataStructureNodes)
+        {
+            List<string> words = GetWords(wordsToAddToNode);
+            List<string> filteredWords = GetFilteredWords(words, DataStructureWordsToIgnore);
+
+            foreach (var word in filteredWords)
             {
-                int nodeIndex = dsWordsForNode.Index;
-                List<string> dsWords = dsWordsForNode.WordsCount.Keys.ToList();
-                dsWords = dsWords.Select(s => s.ToLowerInvariant()).ToList();
-                List<string> dsWordsFiltered = dsWords.Except(dsWordsToIgnore).ToList();
+                AddEdgeToKnowledgeGraph(graph: ref graph, nodeIndex: wordsToAddToNode.Index, nodeToAdd: dataStructureNodes[word]);
+            }
+        }
 
-                foreach (var dsWord in dsWordsFiltered)
+        private static List<string> GetFilteredWords(List<string> words, List<string> wordsToIgnore)
+        {
+            return words.Except(wordsToIgnore).ToList();
+        }
+
+        private static List<string> GetWords(WordCount wordCount)
+        {
+            return wordCount.WordsCount.Keys.Select(word => word.ToLowerInvariant()).ToList();
+        }
+
+        private static List<WordCount> GetDataStructureWordsInEachWikiPage()
+        {
+            return JsonConvert.DeserializeObject<List<WordCount>>(File.ReadAllText("../../../dataStructureWordsCountForNodesInGraph.json"));
+        }
+
+        private static Dictionary<string, KnGNode> GetDataStructureNodesWithoutDuplicates(KnowledgeGraph graph)
+        {
+            List<KnGNode> dataStructureNodesWithDuplicates = graph.KnGraph.Where(n => n.OriginalGraphType == OriginalGraphType.DataStructuresKnGraph).ToList();
+            Dictionary<string, KnGNode> result = new Dictionary<string, KnGNode>();
+
+            foreach (KnGNode node in dataStructureNodesWithDuplicates)
+            {
+                string nodeLabel = node.Label.ToLowerInvariant();
+                if (!result.ContainsKey(nodeLabel))
```

Comment:
Is this if check actually necessary? Dictionaries can't contain duplicate keys. I think if you replace the .add() with a [nodelable] = node, you can remove the if. Unless you're worried about later node values overwriting earlier ones that have the same names

Reply:
yes, I want the first occurrence only. and I can't easily have the dataStructureNodesWithDuplicates have only unique values because there are 2 nodes with the label "Array" but different node indexes. I think the real solution here is to ensure the labels are the unique like the indexes. And this needs to be done when parsing wikipedia.